pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,272 @@
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to read the request body during write", "err", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	firstSuccessfulResponse := make(chan *alertmanagerpb.Response, 1)
+	grpcHeaders := httpTogrpchttpHeaders(r.Header)
+	err = ring.DoBatch(r.Context(), RingOp, d.alertmanagerRing, []uint32{shardByUser(userID)}, func(am ring.InstanceDesc, _ []int) error {
+		d.amSends.WithLabelValues(am.Addr).Inc()
+
+		// Use a background context to make sure all alertmanagers get the request even if we return early.
+		localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.AlertmanagerClient.RemoteTimeout)
+		defer cancel()
+		localCtx = user.InjectOrgID(localCtx, userID)
+		if sp := opentracing.SpanFromContext(r.Context()); sp != nil {
+			localCtx = opentracing.ContextWithSpan(localCtx, sp)
+		}
+		// Get clientIP(s) from Context and add it to localCtx
+		localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+
+		resp, err := d.doRequest(localCtx, am, &httpgrpc.HTTPRequest{
+			Method:  r.Method,
+			Url:     r.RequestURI,
+			Body:    body,
+			Headers: grpcHeaders,
+		})
+		if err != nil {
+			d.amSendFailures.WithLabelValues(am.Addr).Inc()
+			return err
+		}
+
+		if resp.HttpResponse.Code/100 != 2 {
+			return httpgrpc.ErrorFromHTTPResponse(resp.HttpResponse)
+		}
+
+		select {
+		case firstSuccessfulResponse <- resp:
+		default:
+		}
+
+		return nil
+	}, func() {})
+
+	if err != nil {
+		d.respondFromError(err, w, logger)
+	}
+
+	select {
+	case resp := <-firstSuccessfulResponse:
+		d.respondFromGRPCHTTPResponse(w, resp.HttpResponse)
+	default:
+		// This should not happen, but we have this default case to prevent
+		// and bugs deadlocking at this point.
+		level.Warn(logger).Log("msg", "first successful response not available on nil error")
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+func (d *Distributor) doRead(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	key := shardByUser(userID)
+	replicationSet, err := d.alertmanagerRing.Get(key, RingOp, nil, nil, nil)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get replication set from the ring", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	req, err := server.HTTPRequest(r)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get grpc request from http request", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(r.Context(), d.cfg.AlertmanagerClient.RemoteTimeout)
+	defer cancel()
+	// Until we have a mechanism to combine the results from multiple alertmanagers,
+	// we forward the request to only only of the alertmanagers.
+	amDesc := replicationSet.Ingesters[0]
+	d.amSends.WithLabelValues(amDesc.Addr).Inc()
+	resp, err := d.doRequest(ctx, amDesc, req)
+	if err != nil {
+		d.amSendFailures.WithLabelValues(amDesc.Addr).Inc()
```

Comment:
If one fails, should we try another one?

Reply:
Ignore this metric, the "failure" is misleading here, it's removed now. The err is actually the error returned by the AM. The spillover is taken care by the `DoBatch` if an AM is not healthy, it is similar how we are doing for distributor->ingester. I don't think we have to re-try for any error after we have selected AM from the ring.